File: app-miyou/src/main/java/com/miyou/app/infrastructure/monitoring/micrometer/CompositePipelineMetricsReporter.java
package com.miyou.app.infrastructure.monitoring.micrometer;

import java.util.List;

import com.miyou.app.domain.monitoring.model.PipelineSummary;
import com.miyou.app.domain.monitoring.port.PipelineMetricsReporter;

/**
 * 여러 PipelineMetricsReporter를 조합하여 하나의 Reporter로 제공합니다.
 */
public class CompositePipelineMetricsReporter implements PipelineMetricsReporter {

	private final List<PipelineMetricsReporter> reporters;

	public CompositePipelineMetricsReporter(List<PipelineMetricsReporter> reporters) {
		this.reporters = List.copyOf(reporters);
	}

	@Override
	public void report(PipelineSummary summary) {
		for (PipelineMetricsReporter reporter : reporters) {
			reporter.report(summary);
		}
	}
}

File: app-miyou/src/main/kotlin/com/miyou/app/infrastructure/monitoring/micrometer/CompositePipelineMetricsReporter.kt
package com.miyou.app.infrastructure.monitoring.micrometer

import com.miyou.app.domain.monitoring.model.PipelineSummary
import com.miyou.app.domain.monitoring.port.PipelineMetricsReporter

/**
 * 여러 [PipelineMetricsReporter]를 순차적으로 실행해 단일 진입점으로 집계 결과를 기록.
 */
class CompositePipelineMetricsReporter(
    private val reporters: List<PipelineMetricsReporter>,
) : PipelineMetricsReporter {
    init {
        require(reporters.isNotEmpty()) { "reporters must not be empty" }
    }

    override fun report(summary: PipelineSummary) {
        reporters.forEach { reporter ->
            reporter.report(summary)
        }
    }
}
